addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,878 @@
+                    listActions(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("listthermostat")) {
+                    listThermostat(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("listthermostatHVAC")) {
+                    listThermostatHVAC(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("readtariffdata")) {
+                    readTariffData(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("getalarms")) {
+                    getAlarms(nhcCmdAck);
+                } else {
+                    logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+                }
+            } else if (nhcCmdAck.event != null) {
+                if (nhcCmdAck.event.equals("listactions")) {
+                    for (HashMap<String, String> action : nhcCmdAck.data) {
+                        int id = Integer.valueOf(action.get("id"));
+                        setActionState(id, Integer.valueOf(action.get("value1")));
+                    }
+                } else {
+                    logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+                }
+            } else {
+                logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+            }
+        }
+
+    }
+
+    /**
+     * Cleanup socket when the communication with Niko Home Control IP-interface is closed.
+     */
+    private void stopCommunication() {
+
+        listenerStopped = true;
+
+        if (nhcSocket != null) {
+            try {
+                nhcSocket.close();
+            } catch (IOException e) {
+            }
+            nhcSocket = null;
+        }
+        logger.warn("Niko Home Control: communication stopped");
+    }
+
+    /**
+     * Cleanup socket when the communication with Niko Home Control IP-interface is closed.
+     *
+     * @param restart true if it is the intention to try to restart, should not be true if the event listener has not
+     *            successfully been started before
+     */
+    public synchronized void stopCommunication(Boolean restart) {
+        stopCommunication();
+
+        // try restarting the communication
+        if (restart) {
+            logger.info("Niko Home Control: restart communication");
+            startCommunication(nhcAddress, nhcPort.intValue());
+        }
+
+    }
+
+    /**
+     * After setting up the communication with the Niko Home Control IP-interface, send all initialization messages.
+     * <p>
+     * Only at first initialization, also set the return values. Otherwise use the runnable to get updated values.
+     * While communication is set up for thermostats, tariff data and alarms, only info from locations and actions
+     * is used beyond this point in openHAB. All other elements are for future extensions.
+     *
+     * @throws IOException
+     */
+    private void initialize() throws IOException {
+
+        NHCCmd nhcCmd = new NHCCmd();
+
+        nhcCmd.cmd = "systeminfo";
```

Comment:
used a constructor for the more common fields and setters for the additional optional ones.